lib/maths/CBoostedTreeLeafNodeStatistics.cc

```diff
@@ -267,6 +335,20 @@ void CBoostedTreeLeafNodeStatistics::addRowDerivatives(const CEncodedDataFrameRo
 
     auto derivatives = readLossDerivatives(row.unencodedRow(), m_ExtraColumns,
                                            m_NumberLossParameters);
+    if (derivatives.size() == 2) {
+        if (derivatives(0) >= 0.0) {
+            splitsDerivatives.m_PositiveDerivativesGSum += derivatives(0);
+            splitsDerivatives.m_PositiveDerivativesHSum += derivatives(1);
```

Comment:
I would have `m_PositiveDerivatives`, `m_NegativeDerivatives`, `m_PositiveDerivativesMin` and `m_PositiveDerivativesMax` and have `splitsDerivatives.addPositiveDerivatives(derivatives)` and `splitsDerivatives.addNegativeDerivatives(derivatives)`. For example, `addPositiveDerivatives` would look like:
```
  m_PositiveDerivatives += derivatives;
  m_PositiveDerivativesMin.cwiseMin(derivatives);
  m_PositiveDerivativesMax.cwiseMax(derivatives);
```
When you return the value you can do std::max(m_PositiveDerivativesMax(1), 0.0) in place of taking `fabs` here. I would make `m_PositiveDerivatives`, `m_NegativeDerivatives` type Eigen::Matrix<double, 2, 1> no heap allocations. And `m_PositiveDerivativesMin` and `m_PositiveDerivativesMax` type Eigen::Matrix<CFloatStorage, 2, 1>. This way the coefficientwise min and max operations should use float comparison and be vectorised. You just need to initialise `m_PositiveDerivativesMin` to `std::numeric_limits<float>::max()`, etc. The accessors you have have can remain, i.e. you could have a `::negativeDerivativesGSum() const` which reads the value out of the appropriate vector coefficient so the rest of the code doesn't need to now how they're stored and updated.

Reply:
I implemented the suggested change. It doesn't seem to make the code significantly faster, but the implementation is cleaner now.

Reply:
I don't use `CFloatStorage` though to avoid in the hot loop.

Reply:
> I don't use CFloatStorage though to avoid in the hot loop.

That's interesting. `CFloatStorage` will use float comparison operations. Since the underlying types are float it therefore shouldn't coerce them to doubles before comparing. I've definitely found switching min/max operations to use them speed up other code before. Note this was only being proposed for `m_*DerivativesMin` and `m_*DerivativesMax` and not the values you accumulate into.